blacksmith/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+mod channel;
+
+use std::{
+    collections::BTreeMap,
+    fmt::Write,
+    io::{BufRead, BufReader},
+};
+
+use mdbook::{
+    book::{Book, BookItem},
+    errors::Error,
+    preprocess::{Preprocessor, PreprocessorContext}
+};
+
+const CHANNELS: &[&str] = &["stable", "beta", "nightly"];
+const CHANNEL_URL_PREFIX: &str = "https://static.rust-lang.org/dist/channel-rust-";
+const RUSTUP_URLS: &str =
+    "https://raw.githubusercontent.com/rust-lang/rustup.rs/stable/ci/cloudfront-invalidation.txt";
```

Comment:
This is old code AFAICT but looks new in GH, which is the problem I'm trying to address by asking you to split up the commits.

Reply:
The old code is these `const`s, the `channel` module (with the exception of now using `BTreeMap` over `IndexMap`), and line 56–86 of this file. The rest of the code was written by me. The code in main.rs is based on the example preprocessor `mdbook` provides. I don't know if I could split up these parts in a way that perfectly perseveres their history without just refactoring the preprocessor (At the very least everything the `init` function touches) to be entirely within `src/main.rs`, which doesn't seem doesn't seem like an idiomatic solution to me.

Reply:
I think if you just delete all new code from the (last?) commit, commit that -- even though it might not build or work -- and then add a new commit that adds that new code that would work perfectly for me.

Reply:
@Mark-Simulacrum Okay I've made a [test diff](https://github.com/XAMPPRocky/rust-forge/pull/2/commits/6a15d7f1bd57b045c2aa7bdd9883987102b26b45) of what this would like, however I don't think this is as useful as what you were looking for. It seems removing the inline modules creates enough of a diff to not hide the old code.

Reply:
I think that commit is already much more workable to review, so I'll try and schedule some time in to do so today or tomorrow. If I leave comments on old code feel free to just tell me so and ignore them/mark them as resolved without fixing, although if you're feeling up to it opening an issue would be great!